---
title: "RiboseQC vignette"
author: "Lorenzo Calviello"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  rmarkdown::html_vignette:
  html_document:
    toc: true
    toc_float:
      collapsed: false
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{RiboseQC vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup_vig, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

RiboseQC is a package that performs quality control analysis of small RNA-seq data (in .bam format), with a focus on Ribo-seq [1] and related techniques. Thanks to syntax and functions present in packages like *GenomicFeatures*, *rtracklayer* or *BSgenome*, this package can perform comprehensive analyses of alignments on a variety of genomic regions.
This vignette illustrates the usage of the package using example data and annotation from three different experiments performed in three different organisms.

**Warning**: The following analysis is for purposes of demonstrating the package function only. It uses a toy data set composed of the 100 (or fewer) genes in each compartment in the arabidposis data featured in the manuscript - as such the plots appear noiser than in the paper. See the development version on the github for the full version, which downloads the full datasets and is as such unsuitable for a bioconductor package vignette.

As a first step, let's load the package:

```{r load_lib}
library("RiboseQC")

tmp_dir <-  tempdir()


```


## Exploring the annotation

We now create the annotation files for *Arabidopsis thaliana*. We will use a custom gtf file containing annotation of the Araport11 project (https://www.araport.org/), and non-coding RNA species from TAIR10.


```{r arab_1}
arab_fasta <- system.file(package='RiboseQC',"ext_data","simp_arab.fa.gz")
file.copy(arab_fasta,tmp_dir)
system(paste0('gunzip ',file.path(tmp_dir,basename(arab_fasta))))
arab_fasta <- file.path(tmp_dir,gsub(x=basename(arab_fasta),'.gz',''))

stopifnot(is(getSeq(FaFile_Circ(FaFile(arab_fasta))),'DNAStringSet'))

arab_gtf <- system.file(package='RiboseQC',"ext_data","simp_arab.gtf.gz")


```


To parse the rich information present in our *.gtf* file we use the *prepare_annotation_files* function. Such a function creates a *TxDb*  and a compressed *Rdata* file containing several regions of interest and additional information.
The extended vignette makes use of *twobit* files - here we simply make use of a fasta file, which will be converted into a pointer FaFile_Circ object to allow sequence retrieval from circular sequences, and attached ot our annotation object.


```{r create_annot_hum}

annot_file <- prepare_annotation_files(annotation_directory = tmp_dir,
                         genome_seq = arab_fasta,
                         gtf_file = arab_gtf,
                         scientific_name = "arabidopsis.test",
                         annotation_name = "araport11_custom",export_bed_tables_TxDb = TRUE,
                         forge_BSgenome = FALSE,
                         create_TxDb = TRUE)

```


We can now read such information using the *load_annotation* function:

```{r load_hum}

load_annotation(file.path(tmp_dir,"simp_arab.gtf.gz_Rannot"))
genome_seq$circularRanges

```

Two objects have now been created: a *genome_seq* object links to the *BSgenome* package we just created and loaded (containing genome sequences), and a *GTF_annotation* object containing important information present in our *.gtf* file.
For instance, we can access genomic sequences using commands as:

```{r gen_hum}
getSeq(genome_seq,GRanges("Chr1:1-100"))
getSeq(genome_seq,GRanges("ChrM:1-100"))
```

Transcript annotation and CDS annotations can be accessed as follows:

```{r gtf_hum_general_1}
GTF_annotation$exons_txs
```

```{r gtf_hum_general_2}
GTF_annotation$cds_txs
```


The genomic sequences corresponding to such genomic regions can be easily extracted:

```{r gen_hum_cds}
getSeq(genome_seq,GTF_annotation$cds_txs[[4]])
```

A list of annotated start and stop codons, including the transcripts they map to, can be accessed using:

```{r gtf_hum_general_3}
GTF_annotation$start_stop_codons
```

CDS annotation in transcript-level coordinates is also reported:

```{r gtf_hum_general_4}
GTF_annotation$cds_txs_coords
```

A list of gene ids, transcript ids, together with their symbols and biotypes, can be accessed with:

```{r gtf_hum_general_5}
GTF_annotation$trann
```

The genetic codes used for each chromosomes are accessed using:

```{r gtf_hum_general_6}
GTF_annotation$genetic_codes
getGeneticCode(GTF_annotation$genetic_codes["Chr2","genetic_code"])
getGeneticCode(GTF_annotation$genetic_codes["ChrM","genetic_code"])
```

Annotation and genome sequences are linked together in the annotation creation step.
The BSgenome package or FaFile_Circ object corresponding to the .gtf file used is reported in the *GTF_annotation* object:

```{r gtf_hum_general_7}
GTF_annotation$genome
```


##

Let's now download a subset of a Ribo-seq dataset in arabidopsis roots cells.


```{r report_1}
rootbam <-   system.file(package='RiboseQC',"ext_data","simp_arab_root.bam")
```

We will now download two sample datasets from a Ribo-seq experiments in Arabidopsis roots and shoots [5].

We now perform different sets of calculations on our data (including read-length and organelle-specific metagene analyses), using the annotation we previously created. Several files (including automatically generated P_sites positions, and other output from the analysis pipeline) will be automatically created. 
Furthermore, an html dynamic report which illustrated the analysis results is created:

```{r report_2}

annotation=file.path(tmp_dir,"simp_arab.gtf.gz_Rannot")

load_annotation(annotation)

#tmp_dir2="/var/folders/5t/j5mzk77n11b5w4q_54rlq97m0000gn/T//RtmpOFmFiU"

bam_filepath=c(
  system.file(package='RiboseQC',"ext_data","simp_arab_root.bam",mustWork = T),
  system.file(package='RiboseQC',"ext_data","simp_arab_shoots.bam",mustWork = T)
)
```

The full analysis is run like so (this takes some time)

```{r run_analysis,eval = FALSE}

resfile <- RiboseQC_analysis(
  annotation_file=annotation,
  bam_files = bam_filepath,
  fast_mode = TRUE,
#  report_file = file.path(tmp_dir,"test_root_shoots.html"),
  dest_names = file.path(tmp_dir,c("root","shoots")),
  sample_names = c("root","shoots"),
  genome_seq=arab_fasta,
  write_tmp_files = FALSE,
  create_report = FALSE
)

```
Instead we'll use pre-made results files:
```{r load_results}
resfiles=c(
  system.file(package='RiboseQC',"ext_data","root_results_RiboseQC",mustWork = T),
  system.file(package='RiboseQC',"ext_data","shoots_results_RiboseQC",mustWork = T)
)
```

```{r make_report, eval = F}

create_html_report(
  input_files = resfiles,
  output_file = file.path(tmp_dir,"test_root_shoots.html"),
  input_sample_names = c('root','shoots'),
  extended = FALSE,
  knitreport = FALSE
)

```


The html report can be opened by different browsers such as firefox, chrome etc...
Moreover, all the generated plots are available in *.pdf* format and as *RDS* files in the same folder.
Plots can also be generated individually using the results files.
For example, the profile of P-site positions along the CDS can be visualized using:

```{r report_3,warning=FALSE, results='asis', fig.width=12, fig.height=10, dpi=120, eval = TRUE}
res_all <- get(load(resfiles[1]))
dir.create(rec=TRUE,file.path(tmp_dir,"test_root_shoots.html_plots/rds/"))

invisible(plot_metagene_hm_rmd(
  data = res_all,
  profile_type = "profiles_P_sites",
  sample = 'root',
  output_rds_path = file.path(tmp_dir,"test_root_shoots.html_plots/rds/") 
))

readRDS(file.path(tmp_dir,"test_root_shoots.html_plots/rds/root_nucl_4_profiles_P_sites_metagene_subcodon_all"))[[1]]

```


Despite the lack of a dedicated treatment to purify chloroplast ribosomes in these Ribo-seq experiments (e.g. using chloramphenicol), it is possible to detect footprints from distinct organelles, enriched in the shoots sample rather than roots one.

For instance, we can visualize the amount of footprints mapping to different organelles and biotypes in the two samples:

```{r arab_3, eval = TRUE}
res_object_list <- list(
  root = get(load(resfiles[1])),
  shoots = get(load(resfiles[2]))
)
rdsdir <- file.path(tmp_dir,"test_root_shoots.html_plots/rds/")
dir.create( rdsdir , rec=TRUE)
invisible(plot_read_biotype_dist_2(rdata_list = res_object_list, output_rds_path = rdsdir))
readRDS(file.path(tmp_dir,"test_root_shoots.html_plots/rds/all_samples_1_readlocdist"))[[1]]

```

or visualize z-scored positional codon usage values, calculated using A-site positions for chloroplast ribosomes in the shoots sample:

```{r arab_4, warning=FALSE, results='asis', fig.width=12, fig.height=16, dpi=150, eval = FALSE}

plot_codon_usage_positional_rmd(res_object_list[[2]], 'shoots',output_rds_path =  rdsdir)


```

The report allows to compare different samples side-by-side using tabs for different samples.
For example, the 5'end profiles of different read lengths around start and stop codons can be visualized for different experiments. For the single plots:


```{r yeast_4, warning=FALSE, results='asis', fig.width=12, fig.height=10, dpi=120,eval= FALSE}

invisible(plot_metagene_hm_rmd(
  data = res_all,
  profile_type = "profiles_P_sites",
  sample = 'shoots',
  output_rds_path = file.path(tmp_dir,"test_root_shoots.html_plots/rds/") 
))

readRDS(file.path(tmp_dir,"test_root_shoots.html_plots/rds/root_nucl_4_profiles_P_sites_metagene_subcodon_all"))[[1]]

invisible(plot_metagene_hm_rmd(
  data = res_all,
  profile_type = "profiles_fivepr",
  sample = 'root',
  output_rds_path = file.path(tmp_dir,"test_root_shoots.html_plots/rds/") 
))


as_ggplot(readRDS(file.path(tmp_dir,"test_root_shoots.html_plots/rds/root_nucl_4_profiles_fivepr_metagene_subcodon_log2"))[[1]])

invisible(plot_metagene_hm_rmd(
  data = res_all,
  profile_type = "profiles_fivepr",
  sample = 'shoots',
  output_rds_path = file.path(tmp_dir,"test_root_shoots.html_plots/rds/") 
))

as_ggplot(readRDS(file.path(tmp_dir,"test_root_shoots.html_plots/rds/shoots_nucl_4_profiles_fivepr_metagene_subcodon_log2"))[[1]])

```


For a more extended vignette, please see the package github .

Much more information is available in the html reports, including read length distributions per biotype and compartment, statistics on codon usage, and analysis of the highest mapping positions in the genome.

## References

1) Ingolia, N.T. et al. (2009) Genome-wide analysis in vivo of resolution using ribosome profiling. Science 324, 218–223

2) Calviello, L. et al. (2015) Detecting actively translated open reading frames in ribosome profiling data. Nat. Methods 13, 1–9

3) Daechan P. et al. (2014) Simultaneous mapping of transcript ends at single-nucleotide resolution and identification of widespread promoter-associated non-coding RNA governed by TATA elements, Nucleic Acids Research, Volume 42, Issue 6, 1 April 2014, Pages 3736-3749

4) Archer, S.K. et al. (2016) Dynamics of ribosome scanning and recycling revealed by translation complex profiling. Nature 535, 570–574

5) Hsu, P.Y. et al. (2016) Super-resolution ribosome profiling reveals novel translation events in Arabidopsis. PNAS, 113, E7126–E7135


## Session info

```{r end_sessinf}
session_info()
```



