---
title: "RiboseQC Analysis"
author: "Dermo Harnett"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{RiboseQC Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(RiboseQC)
```

## Step 1. Creating Annotation and Sequence Files for RiboseQC

The first step is to create the necessary bioconductor objects and store them in a folder so they can be re-used. The function ```prepare_annotation_file``` takes the name of the directory it should store the annoation, as well as a gtf file and a twobit file containing sequence information for your data. You can also supply a scientific name for the organism, and a name for the gene annotation, for use in the metadata of the BSgenome package that will be created.

```{r, fig.show='hold'}
gtf_file <- system.file("extdata", "example.gtf", 
                        package = "RiboseQC",mustWork = TRUE)
twobit_file <- system.file("extdata", "example.2bit",
                           package = "RiboseQC",mustWork = TRUE)
output_dir <- tempdir()

prepare_annotation_files(output_dir,twobit_file,gtf_file,
                         scientific_name = "Arabidopsis.thaliana", 
                         annotation_name = "tair10.42" )

annotation_file <- file.path(output_dir,paste0(basename(gtf_file),'_Rannot'))

stopifnot(file.exists(annotation_file))

output_dir%>%list.files

```


## Step 2: Runnning RiboseQC

RiboseQC can be run with a single command. Riboseqc_analysis takes as it's input the annotation file created in the previous step, and a vector of bam files. Options allow tweaking of the procedure or p-site selection and which read lengths are included for downstream analysis. 

```{r, echo=FALSE, results='asis'}

bam_file <- system.file('extdata','example.bam',
                        package='RiboseQC',mustWork = TRUE)

RiboseQC_analysis(annotation_file = annotation_file,
                  bam_files = bam_file,
                  fast_mode = TRUE,
                  create_report = FALSE )


file.exists(bam_file)

dirname(bam_file)%>%list.files

```
